src/lib/shell/MainLoopZephyr.cpp

```diff
@@ -14,30 +14,117 @@
  *    See the License for the specific language governing permissions and
  *    limitations under the License.
  */
+#include <cstring>
 #include <init.h>
 #include <shell/shell.h>
+#include <stdio.h>
 
 #include <lib/core/CHIPError.h>
 #include <lib/shell/Engine.h>
 #include <lib/shell/streamer_zephyr.h>
+#include <lib/support/CodeUtils.h>
 
+using chip::Shell::cmd_completion_context;
 using chip::Shell::Engine;
+using chip::Shell::shell_command_t;
+using chip::Shell::shell_map_t;
 
 static int cmd_matter(const struct shell * shell, size_t argc, char ** argv)
 {
     chip::Shell::streamer_set_shell(shell);
     return (Engine::Root().ExecCommand(argc - 1, argv + 1) == CHIP_NO_ERROR) ? 0 : -ENOEXEC;
 }
 
+/**
+ * This recursive function obtains all sub-commands (possbile completions) for a command node
+ * and build it into a tree of Zephyr's shell_static_entry structure.
+ *
+ * @param prefix                  The path/prefix until the root command node of the tree, without tailing space.
+ *                                e.g. "dns browse". Set to emptry string "" if building tree for a root-level command.
+ * @param syntax                  The single-word command name/syntax for the root node of the tree. e.g. "resolve".
+ *                                Set to emptry string "" if calling to build from the root (to include all commands).
+ * @param help                    The help text for the command at roote node of the tree. Set to NULL if none.
+ *
+ * @return                        The shell_static_entry structure tree that chains all the sub-commands from the
+ *                                requested command node.
+ *
+ */
+const struct shell_static_entry * build_command_tree(const char * prefix, const char * syntax, const char * help)
+{
+
+    char * sub_prefix = new char[strlen(prefix) + strlen(syntax) + 2];
+    size_t pos        = 0;
+    if (strcmp(prefix, "") != 0)
+    {
+        strncpy(&sub_prefix[pos], prefix, strlen(prefix));
+        pos += strlen(prefix);
+        strncpy(&sub_prefix[pos], " ", 1);
+        pos += 1;
+    }
+    strncpy(&sub_prefix[pos], syntax, strlen(syntax) + 1);
+
+    cmd_completion_context context = cmd_completion_context(sub_prefix);
+    CHIP_ERROR ret                 = Engine::GetCommandCompletions(&context);
+
+    const struct shell_static_entry * static_entry;
+    if (ret == CHIP_NO_ERROR && context.cmdc != 0)
+    {
+        const struct shell_static_entry * sub_static_entry = new shell_static_entry[context.cmdc + 1];
```

Comment:
Add comment about the lifetime of this as it does not seem to have a corresponding delete[].